2024-01-28 11:32:00,266: root: INFO: MLTrader : Executing the initialize lifecycle method
2024-01-28 11:32:00,642: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:00,684: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:01,332: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:01,396: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:01,623: root: INFO: Current backtesting datetime 2023-12-15 08:30:00-05:00
2024-01-28 11:32:01,627: root: INFO: MLTrader : Executing the before_market_opens lifecycle method
2024-01-28 11:32:01,628: root: INFO: Current backtesting datetime 2023-12-15 09:30:00-05:00
2024-01-28 11:32:01,628: root: INFO: MLTrader : Executing the before_starting_trading lifecycle method
2024-01-28 11:32:01,629: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:01[0m
2024-01-28 11:32:01,630: root: INFO: [32mNew market order of | 10 SPY buy | with status unprocessed was submitted.[0m
2024-01-28 11:32:01,633: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:01[0m
2024-01-28 11:32:01,633: root: INFO: Getting historical prices for SPY, 1 bars, 
2024-01-28 11:32:01,634: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,036: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:02,049: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:02,514: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,515: root: INFO: [32mFilled Transaction: buy 10 of SPY at 469.48999023 USD per share[0m
2024-01-28 11:32:02,515: root: INFO: market order of | 10 SPY buy | with status new was filled
2024-01-28 11:32:02,521: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,525: root: INFO: Current backtesting datetime 2023-12-15 16:00:00-05:00
2024-01-28 11:32:02,525: root: INFO: Current backtesting datetime 2023-12-16 09:30:00-05:00
2024-01-28 11:32:02,528: root: INFO: [34mMLTrader : The market is not currently open, skipping this trading iteration[0m
2024-01-28 11:32:02,530: root: INFO: Current backtesting datetime 2023-12-16 09:29:00-05:00
2024-01-28 11:32:02,532: root: INFO: Current backtesting datetime 2023-12-16 09:29:00-05:00
2024-01-28 11:32:02,532: root: INFO: MLTrader : Executing the after_market_closes lifecycle method
2024-01-28 11:32:02,533: root: INFO: Current backtesting datetime 2023-12-18 08:30:00-05:00
2024-01-28 11:32:02,533: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,535: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,536: root: INFO: MLTrader : Executing the before_market_opens lifecycle method
2024-01-28 11:32:02,538: root: INFO: Current backtesting datetime 2023-12-18 09:30:00-05:00
2024-01-28 11:32:02,538: root: INFO: MLTrader : Executing the before_starting_trading lifecycle method
2024-01-28 11:32:02,542: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,542: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,543: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,544: root: INFO: Current backtesting datetime 2023-12-18 16:00:00-05:00
2024-01-28 11:32:02,544: root: INFO: Current backtesting datetime 2023-12-19 09:30:00-05:00
2024-01-28 11:32:02,546: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,546: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,547: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,548: root: INFO: Current backtesting datetime 2023-12-19 16:00:00-05:00
2024-01-28 11:32:02,548: root: INFO: Current backtesting datetime 2023-12-20 09:30:00-05:00
2024-01-28 11:32:02,549: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,549: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,550: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,551: root: INFO: Current backtesting datetime 2023-12-20 16:00:00-05:00
2024-01-28 11:32:02,551: root: INFO: Current backtesting datetime 2023-12-21 09:30:00-05:00
2024-01-28 11:32:02,553: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,553: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,554: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,555: root: INFO: Current backtesting datetime 2023-12-21 16:00:00-05:00
2024-01-28 11:32:02,556: root: INFO: Current backtesting datetime 2023-12-22 09:30:00-05:00
2024-01-28 11:32:02,557: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,558: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,559: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,559: root: INFO: Current backtesting datetime 2023-12-22 16:00:00-05:00
2024-01-28 11:32:02,560: root: INFO: Current backtesting datetime 2023-12-23 09:30:00-05:00
2024-01-28 11:32:02,561: root: INFO: [34mMLTrader : The market is not currently open, skipping this trading iteration[0m
2024-01-28 11:32:02,563: root: INFO: Current backtesting datetime 2023-12-23 09:29:00-05:00
2024-01-28 11:32:02,565: root: INFO: Current backtesting datetime 2023-12-23 09:29:00-05:00
2024-01-28 11:32:02,565: root: INFO: MLTrader : Executing the after_market_closes lifecycle method
2024-01-28 11:32:02,566: root: INFO: Current backtesting datetime 2023-12-26 08:30:00-05:00
2024-01-28 11:32:02,566: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,567: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,568: root: INFO: MLTrader : Executing the before_market_opens lifecycle method
2024-01-28 11:32:02,569: root: INFO: Current backtesting datetime 2023-12-26 09:30:00-05:00
2024-01-28 11:32:02,569: root: INFO: MLTrader : Executing the before_starting_trading lifecycle method
2024-01-28 11:32:02,572: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,573: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,574: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,574: root: INFO: Current backtesting datetime 2023-12-26 16:00:00-05:00
2024-01-28 11:32:02,575: root: INFO: Current backtesting datetime 2023-12-27 09:30:00-05:00
2024-01-28 11:32:02,576: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,576: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,577: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,578: root: INFO: Current backtesting datetime 2023-12-27 16:00:00-05:00
2024-01-28 11:32:02,578: root: INFO: Current backtesting datetime 2023-12-28 09:30:00-05:00
2024-01-28 11:32:02,579: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,579: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,580: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,581: root: INFO: Current backtesting datetime 2023-12-28 16:00:00-05:00
2024-01-28 11:32:02,581: root: INFO: Current backtesting datetime 2023-12-29 09:30:00-05:00
2024-01-28 11:32:02,583: root: INFO: [34mMLTrader : Executing the on_trading_iteration lifecycle method at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,583: root: INFO: [34mMLTrader : Trading iteration ended at 2024-01-28 11:32:02[0m
2024-01-28 11:32:02,583: root: INFO: MLTrader : [34mSleeping for 86400 seconds[0m
2024-01-28 11:32:02,584: root: INFO: Current backtesting datetime 2023-12-29 16:00:00-05:00
2024-01-28 11:32:02,584: root: INFO: Current backtesting datetime 2023-12-30 09:30:00-05:00
2024-01-28 11:32:02,586: root: INFO: [34mMLTrader : The market is not currently open, skipping this trading iteration[0m
2024-01-28 11:32:02,587: root: INFO: Current backtesting datetime 2023-12-30 09:29:00-05:00
2024-01-28 11:32:02,590: root: INFO: Current backtesting datetime 2023-12-30 09:29:00-05:00
2024-01-28 11:32:02,590: root: INFO: MLTrader : Executing the after_market_closes lifecycle method
2024-01-28 11:32:02,591: root: INFO: Current backtesting datetime 2024-01-02 08:30:00-05:00
2024-01-28 11:32:02,591: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,592: root: WARNING: quote is not implemented for YahooData, but USD was passed as the quote
2024-01-28 11:32:02,593: root: INFO: MLTrader : Executing the before_market_opens lifecycle method
2024-01-28 11:32:02,595: root: INFO: Current backtesting datetime 2024-01-02 09:30:00-05:00
2024-01-28 11:32:02,595: root: INFO: MLTrader : Executing the before_starting_trading lifecycle method
2024-01-28 11:32:02,598: root: INFO: Current backtesting datetime 2024-01-02 15:59:00-05:00
2024-01-28 11:32:02,599: root: INFO: MLTrader : Executing the before_market_closes lifecycle method
2024-01-28 11:32:02,601: root: INFO: Current backtesting datetime 2024-01-02 16:00:00-05:00
2024-01-28 11:32:02,601: root: INFO: MLTrader : Executing the after_market_closes lifecycle method
2024-01-28 11:32:02,601: root: INFO: MLTrader : Executing the on_strategy_end lifecycle method
2024-01-28 11:32:02,937: root: INFO: Backtesting finished
2024-01-28 11:32:03,789: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:03,801: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:04,311: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:04,324: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:04,928: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:04,959: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:05,461: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:05,493: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:06,521: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:06,560: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:07,170: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:07,185: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:07,685: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:07,700: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:08,297: alpaca.trading.stream: ERROR: error during websocket communication: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 170, in _run_forever
    await self._start_ws()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 130, in _start_ws
    await self._connect()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/alpaca/trading/stream.py", line 61, in _connect
    self._ws = await websockets.connect(self._endpoint, **self._websocket_params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
    return await self.__await_impl__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl__
    _transport, _protocol = await self._create_connection()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1112, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 1145, in _create_connection_transport
    await waiter
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 574, in _on_handshake_complete
    raise handshake_exc
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/sslproto.py", line 556, in _do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)
2024-01-28 11:32:08,310: alpaca.trading.stream: INFO: starting trading websocket connection
2024-01-28 11:32:08,487: backtest_stats: INFO: Backtest took 0:00:08.226135 for a speed of 0.000
